File: app/src/main/java/com/sjn/stamp/ui/custom/PeriodSelectLayout.java
package com.sjn.stamp.ui.custom;

import android.app.DatePickerDialog;
import android.content.Context;
import android.content.res.Resources;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;

import com.sjn.stamp.R;
import com.sjn.stamp.utils.TimeHelper;

import org.joda.time.LocalDate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

public class PeriodSelectLayout extends LinearLayout {

    public static class Period {
        PeriodType mPeriodType;
        LocalDate mFrom;
        LocalDate mTo;

        public static Period latestWeek() {
            return new Period(PeriodType.CUSTOM, TimeHelper.getJapanToday().minusWeeks(1), TimeHelper.getJapanToday());
        }

        Period(PeriodType periodType, LocalDate from, LocalDate to) {
            mPeriodType = periodType;
            mFrom = from;
            mTo = to;
        }

        public String toString(Resources resources) {
            switch (mPeriodType) {
                case TOTAL:
                    return resources.getString(R.string.period_label_total);
                case YEARLY:
                    return resources.getString(R.string.period_label_year, String.valueOf(mTo.getYear()));
                case MONTHLY:
                    return resources.getString(R.string.period_label_month, String.valueOf(mTo.getYear()), String.valueOf(mTo.getMonthOfYear()));
                case DAIRY:
                    return resources.getString(R.string.period_label_day, String.valueOf(mTo.getYear()), String.valueOf(mTo.getMonthOfYear()), String.valueOf(mTo.getDayOfMonth()));
                case CUSTOM:
                    return mFrom.toString() + "~" + mTo.toString();
                default:
                    return "";
            }
        }

        public PeriodType getPeriodType() {
            return mPeriodType;
        }

        public Date from() {
            if (getPeriodType() == PeriodType.TOTAL) {
                return null;
            }
            if (mFrom == null) {
                return null;
            }
            return mFrom.toDateTimeAtStartOfDay().toDate();
        }


        public Date to() {
            if (getPeriodType() == PeriodType.TOTAL) {
                return null;
            }
            if (mTo == null) {
                return null;
            }
            return mTo.toDateTimeAtStartOfDay().plusDays(1).toDate();
        }
    }

    interface PeriodChangeListener {
        void onChange(Period period);
    }

    public enum PeriodType {
        TOTAL(0, R.string.period_total, GONE, GONE, GONE, GONE),
        YEARLY(1, R.string.period_yearly, VISIBLE, GONE, GONE, GONE),
        MONTHLY(2, R.string.period_monthly, VISIBLE, VISIBLE, GONE, GONE),
        DAIRY(3, R.string.period_dairy, VISIBLE, VISIBLE, VISIBLE, GONE),
        CUSTOM(4, R.string.period_custom, GONE, GONE, GONE, VISIBLE);

        final public int mValue;
        final public int mTextId;
        final public int mYearVisibility;
        final public int mMonthVisibility;
        final public int mDayVisibility;
        final public int mCustomVisibility;

        PeriodType(int value, int textId, int yearVisibility, int monthVisibility, int dayVisibility, int customVisibility) {
            mValue = value;
            mTextId = textId;
            mYearVisibility = yearVisibility;
            mMonthVisibility = monthVisibility;
            mDayVisibility = dayVisibility;
            mCustomVisibility = customVisibility;
        }

        public static PeriodType of(int value) {
            for (PeriodType periodType : PeriodType.values()) {
                if (periodType.mValue == value) return periodType;
            }
            return null;
        }

        public static String[] strings(Resources resources) {
            List<String> strings = new ArrayList<>();
            for (PeriodType periodType : PeriodType.values()) {
                strings.add(resources.getString(periodType.mTextId));
            }
            return strings.toArray(new String[0]);
        }

        public int getValue() {
            return mValue;
        }
    }

    Period mPeriod;
    View mLayout;
    ViewGroup mYearLayout;
    ViewGroup mMonthLayout;
    ViewGroup mDayLayout;
    ViewGroup mCustomLayout;
    EditText mDatePickerFrom;
    EditText mDatePickerTo;
    Spinner mPeriodTypeSpinner;
    Spinner mSpinnerDay;
    Spinner mSpinnerMonth;
    Spinner mSpinnerYear;
    PeriodChangeListener mPeriodChangeListener;

    public PeriodSelectLayout(Context context) {
        super(context);
    }

    public PeriodSelectLayout(Context context, AttributeSet attrs) {
        this(context, attrs, Period.latestWeek());
    }

    public PeriodSelectLayout(Context context, AttributeSet attr, Period defaultPeriod) {
        super(context, attr);
        mPeriod = defaultPeriod;
        mLayout = LayoutInflater.from(context).inflate(getResourceId(), this);
        mYearLayout = mLayout.findViewById(R.id.period_year);
        mMonthLayout = mLayout.findViewById(R.id.period_month);
        mDayLayout = mLayout.findViewById(R.id.period_day);
        mCustomLayout = mLayout.findViewById(R.id.period_custom);
        mPeriodTypeSpinner = mLayout.findViewById(R.id.period_kind_spinner);
        mSpinnerDay = mLayout.findViewById(R.id.period_day_spinner);
        mSpinnerMonth = mLayout.findViewById(R.id.period_month_spinner);
        mSpinnerYear = mLayout.findViewById(R.id.period_year_spinner);
        mDatePickerFrom = mLayout.findViewById(R.id.date_picker_from);
        mDatePickerFrom.setText(mPeriod.mFrom.toString());
        mDatePickerFrom.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                final DatePickerDialog datePickerDialog = new DatePickerDialog(
                        getContext(),
                        new DatePickerDialog.OnDateSetListener() {
                            @Override
                            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                                mPeriod.mFrom = new LocalDate(year, monthOfYear + 1, dayOfMonth);
                                onPeriodChange();
                            }
                        },
                        mPeriod.mFrom.getYear(), mPeriod.mFrom.getMonthOfYear() - 1, mPeriod.mFrom.getDayOfMonth());
                datePickerDialog.show();
            }
        });
        mDatePickerTo = mLayout.findViewById(R.id.date_picker_to);
        mDatePickerTo.setText(mPeriod.mTo.toString());
        mDatePickerTo.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                final DatePickerDialog datePickerDialog = new DatePickerDialog(
                        getContext(),
                        new DatePickerDialog.OnDateSetListener() {
                            @Override
                            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                                mPeriod.mTo = new LocalDate(year, monthOfYear + 1, dayOfMonth);
                                onPeriodChange();
                            }
                        },
                        mPeriod.mTo.getYear(), mPeriod.mTo.getMonthOfYear() - 1, mPeriod.mTo.getDayOfMonth());
                datePickerDialog.show();
            }
        });
        //noinspection ResourceType
        mYearLayout.setVisibility(mPeriod.mPeriodType.mYearVisibility);
        //noinspection ResourceType
        mMonthLayout.setVisibility(mPeriod.mPeriodType.mMonthVisibility);
        //noinspection ResourceType
        mDayLayout.setVisibility(mPeriod.mPeriodType.mDayVisibility);
        //noinspection ResourceType
        mCustomLayout.setVisibility(mPeriod.mPeriodType.mCustomVisibility);
        initPeriodTypeSpinner();
        initDaySpinner();
        initMonthSpinner();
        initYearSpinner();
    }

    int getResourceId() {
        return R.layout.layout_period_select;
    }

    public Period getPeriod() {
        return mPeriod;
    }

    void initPeriodTypeSpinner() {
        mPeriodTypeSpinner.setAdapter(createSpinnerAdapter(PeriodType.strings(getResources())));
        mPeriodTypeSpinner.setSelection(mPeriod.mPeriodType.getValue());
        mPeriodTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (mPeriodTypeSpinner.getVisibility() != VISIBLE) {
                    return;
                }
                if (PeriodType.of(i) == null) {
                    return;
                }
                mPeriod.mPeriodType = PeriodType.of(i);
                //noinspection ResourceType
                mYearLayout.setVisibility(mPeriod.mPeriodType.mYearVisibility);
                //noinspection ResourceType
                mMonthLayout.setVisibility(mPeriod.mPeriodType.mMonthVisibility);
                //noinspection ResourceType
                mDayLayout.setVisibility(mPeriod.mPeriodType.mDayVisibility);
                //noinspection ResourceType
                mCustomLayout.setVisibility(mPeriod.mPeriodType.mCustomVisibility);
                updatePeriodBySpinner();
                onPeriodChange();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

    }

    void initDaySpinner() {
        Integer[] days = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31};
        mSpinnerDay.setAdapter(createSpinnerAdapter(days));
        mSpinnerDay.setSelection(findOr0(days, TimeHelper.getJapanNow().getDayOfMonth()));
        mSpinnerDay.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                updatePeriodBySpinner();
                onPeriodChange();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    void initMonthSpinner() {
        Integer[] months = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
        mSpinnerMonth.setAdapter(createSpinnerAdapter(months));
        mSpinnerMonth.setSelection(findOr0(months, TimeHelper.getJapanNow().getMonthOfYear()));
        mSpinnerMonth.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                List<Integer> days = new ArrayList<>();
                int maxDay = mPeriod.mTo.withMonthOfYear(Integer.valueOf(mSpinnerMonth.getSelectedItem().toString())).dayOfMonth().withMaximumValue().getDayOfMonth();
                for (int day = 1; day <= maxDay; day++) {
                    days.add(day);
                }
                mSpinnerDay.setAdapter(createSpinnerAdapter(days.toArray()));
                mSpinnerDay.setSelection(findOr0(days.toArray(new Integer[days.size()]), TimeHelper.getJapanNow().getDayOfMonth()));
                updatePeriodBySpinner();
                onPeriodChange();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    void initYearSpinner() {
        Integer[] years = {2017, 2018, 2019, 2020, 2021, 2022, 2023, 2024, 2025, 2026, 2027, 2028, 2029, 2030};
        mSpinnerYear.setAdapter(createSpinnerAdapter(years));
        mSpinnerYear.setSelection(findOr0(years, TimeHelper.getJapanNow().getYear()));
        mSpinnerYear.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                updatePeriodBySpinner();
                onPeriodChange();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    void onPeriodChange() {
        mDatePickerFrom.setText(mPeriod.mFrom.toString());
        mDatePickerTo.setText(mPeriod.mTo.toString());
        if (mPeriodChangeListener != null) {
            mPeriodChangeListener.onChange(mPeriod);
        }
    }

    void updatePeriodBySpinner() {
        if (mYearLayout.getVisibility() == VISIBLE) {
            mPeriod.mFrom = mPeriod.mFrom.withYear(Integer.valueOf(mSpinnerYear.getSelectedItem().toString())).dayOfYear().withMinimumValue();
            mPeriod.mTo = mPeriod.mTo.withYear(Integer.valueOf(mSpinnerYear.getSelectedItem().toString())).dayOfYear().withMaximumValue();
        }
        if (mMonthLayout.getVisibility() == VISIBLE) {
            mPeriod.mFrom = mPeriod.mFrom.withMonthOfYear(Integer.valueOf(mSpinnerMonth.getSelectedItem().toString())).dayOfMonth().withMinimumValue();
            mPeriod.mTo = mPeriod.mTo.withMonthOfYear(Integer.valueOf(mSpinnerMonth.getSelectedItem().toString())).dayOfMonth().withMaximumValue();
        }
        if (mDayLayout.getVisibility() == VISIBLE) {
            mPeriod.mFrom = mPeriod.mFrom.withDayOfMonth(Integer.valueOf(mSpinnerDay.getSelectedItem().toString()));
            mPeriod.mTo = mPeriod.mTo.withDayOfMonth(Integer.valueOf(mSpinnerDay.getSelectedItem().toString()));
        }
    }

    int findOr0(Integer[] array, int value) {
        return Arrays.asList(array).contains(value) ? Arrays.asList(array).indexOf(value) : 0;
    }

    ArrayAdapter createSpinnerAdapter(Object[] strings) {
        ArrayAdapter spinnerArrayAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, strings);
        spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        return spinnerArrayAdapter;
    }

}

File: app/src/main/java/com/sjn/stamp/ui/custom/PeriodSelectLayout.kt
package com.sjn.stamp.ui.custom

import android.app.DatePickerDialog
import android.content.Context
import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.sjn.stamp.R
import com.sjn.stamp.utils.TimeHelper
import org.joda.time.LocalDate
import java.util.*

class PeriodSelectLayout(context: Context, var period: Period) : LinearLayout(context) {

    private var yearLayout: ViewGroup
    private var monthLayout: ViewGroup
    private var dayLayout: ViewGroup
    private var customLayout: ViewGroup
    private var datePickerFrom: EditText
    private var datePickerTo: EditText
    private var periodTypeSpinner: Spinner
    private var spinnerDay: Spinner
    private var spinnerMonth: Spinner
    private var spinnerYear: Spinner
    private var periodChangeListener: PeriodChangeListener? = null

    interface PeriodChangeListener {
        fun onChange(period: Period)
    }

    class Period internal constructor(var periodType: PeriodType, internal var from: LocalDate?, internal var to: LocalDate?) {

        fun toString(resources: Resources): String = periodType.toString(resources, from, to)
        fun from(): Date? = if (periodType == PeriodType.TOTAL) null else from?.toDateTimeAtStartOfDay()?.toDate()
        fun to(): Date? = if (periodType == PeriodType.TOTAL) null else to?.toDateTimeAtStartOfDay()?.plusDays(1)?.toDate()

        companion object {
            fun latestWeek(): Period = Period(PeriodType.CUSTOM, TimeHelper.japanToday.minusWeeks(1), TimeHelper.japanToday)
        }
    }

    enum class PeriodType(val value: Int, val textId: Int, val yearVisibility: Int, val monthVisibility: Int, val dayVisibility: Int, val customVisibility: Int) {
        TOTAL(0, R.string.period_total, View.GONE, View.GONE, View.GONE, View.GONE) {
            override fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String =
                    resources.getString(R.string.period_label_total)
        },
        YEARLY(1, R.string.period_yearly, View.VISIBLE, View.GONE, View.GONE, View.GONE) {
            override fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String =
                    resources.getString(R.string.period_label_year, to?.year.toString())
        },
        MONTHLY(2, R.string.period_monthly, View.VISIBLE, View.VISIBLE, View.GONE, View.GONE) {
            override fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String =
                    resources.getString(R.string.period_label_month, to?.year.toString(), to?.monthOfYear.toString())
        },
        DAIRY(3, R.string.period_dairy, View.VISIBLE, View.VISIBLE, View.VISIBLE, View.GONE) {
            override fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String =
                    resources.getString(R.string.period_label_day, to?.year.toString(), to?.monthOfYear.toString(), to?.dayOfMonth.toString())
        },
        CUSTOM(4, R.string.period_custom, View.GONE, View.GONE, View.GONE, View.VISIBLE) {
            override fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String =
                    from?.toString() + "~" + to?.toString()
        };

        companion object {
            fun of(value: Int): PeriodType? = PeriodType.values().firstOrNull { it.value == value }
            fun strings(resources: Resources): Array<String> = PeriodType.values().map { resources.getString(it.textId) }.toTypedArray()
        }

        abstract fun toString(resources: Resources, from: LocalDate?, to: LocalDate?): String
    }

    init {
        val layout = LayoutInflater.from(context).inflate(R.layout.layout_period_select, this)
        yearLayout = layout.findViewById(R.id.period_year)
        monthLayout = layout.findViewById(R.id.period_month)
        dayLayout = layout.findViewById(R.id.period_day)
        customLayout = layout.findViewById(R.id.period_custom)
        periodTypeSpinner = layout.findViewById(R.id.period_kind_spinner)
        spinnerDay = layout.findViewById(R.id.period_day_spinner)
        spinnerMonth = layout.findViewById(R.id.period_month_spinner)
        spinnerYear = layout.findViewById(R.id.period_year_spinner)
        datePickerFrom = layout.findViewById(R.id.date_picker_from)
        datePickerTo = layout.findViewById(R.id.date_picker_to)
        initPeriodTypeSpinner()
        initDaySpinner()
        initMonthSpinner()
        initYearSpinner()
        period.from?.let { from ->
            datePickerFrom.setText(from.toString())
            datePickerFrom.setOnClickListener {
                DatePickerDialog(
                        getContext(),
                        DatePickerDialog.OnDateSetListener { _, year, monthOfYear, dayOfMonth ->
                            period.from = LocalDate(year, monthOfYear + 1, dayOfMonth)
                            onPeriodChange()
                        },
                        from.year, from.monthOfYear - 1, from.dayOfMonth).show()
            }
        }
        period.to?.let { to ->
            datePickerTo.setText(to.toString())
            datePickerTo.setOnClickListener {
                DatePickerDialog(
                        getContext(),
                        DatePickerDialog.OnDateSetListener { _, year, monthOfYear, dayOfMonth ->
                            period.to = LocalDate(year, monthOfYear + 1, dayOfMonth)
                            onPeriodChange()
                        },
                        to.year, to.monthOfYear - 1, to.dayOfMonth).show()
            }
        }
        reloadVisibilities()
    }

    constructor(context: Context) : this(context, Period.latestWeek())

    private fun reloadVisibilities() {
        yearLayout.visibility = period.periodType.yearVisibility
        monthLayout.visibility = period.periodType.monthVisibility
        dayLayout.visibility = period.periodType.dayVisibility
        customLayout.visibility = period.periodType.customVisibility
    }

    private fun initPeriodTypeSpinner() {
        periodTypeSpinner.adapter = createSpinnerAdapter(PeriodType.strings(resources))
        periodTypeSpinner.setSelection(period.periodType.value)
        periodTypeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                if (periodTypeSpinner.visibility != View.VISIBLE) {
                    return
                }
                val periodType = PeriodType.of(i) ?: return
                period.periodType = periodType
                reloadVisibilities()
                updatePeriodBySpinner()
                onPeriodChange()
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {

            }
        }

    }

    private fun initDaySpinner() {
        val days = arrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31)
        spinnerDay.adapter = createSpinnerAdapter(days)
        spinnerDay.setSelection(findOr0(days, TimeHelper.japanNow.dayOfMonth))
        spinnerDay.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                updatePeriodBySpinner()
                onPeriodChange()
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {

            }
        }
    }

    private fun initMonthSpinner() {
        val months = arrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
        spinnerMonth.adapter = createSpinnerAdapter(months)
        spinnerMonth.setSelection(findOr0(months, TimeHelper.japanNow.monthOfYear))
        spinnerMonth.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                val maxDay = period.to?.withMonthOfYear(Integer.valueOf(spinnerMonth.selectedItem.toString()))?.dayOfMonth()?.withMaximumValue()?.dayOfMonth
                        ?: return
                val days = (1..maxDay).toList()
                spinnerDay.adapter = createSpinnerAdapter(days.toTypedArray())
                spinnerDay.setSelection(findOr0(days.toTypedArray(), TimeHelper.japanNow.dayOfMonth))
                updatePeriodBySpinner()
                onPeriodChange()
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {

            }
        }
    }

    private fun initYearSpinner() {
        var year = 2017
        val years = mutableListOf<Int>()
        while (year <= TimeHelper.japanNow.year) {
            years.add(year)
            year++
        }
        spinnerYear.adapter = createSpinnerAdapter(years.toTypedArray())
        spinnerYear.setSelection(findOr0(years.toTypedArray(), TimeHelper.japanNow.year))
        spinnerYear.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                updatePeriodBySpinner()
                onPeriodChange()
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {

            }
        }
    }

    private fun onPeriodChange() {
        datePickerFrom.setText(period.from?.toString())
        datePickerTo.setText(period.to?.toString())
        periodChangeListener?.onChange(period)
    }

    private fun updatePeriodBySpinner() {
        if (yearLayout.visibility == View.VISIBLE) {
            period.from = period.from?.withYear(Integer.valueOf(spinnerYear.selectedItem.toString()))?.dayOfYear()?.withMinimumValue()
            period.to = period.to?.withYear(Integer.valueOf(spinnerYear.selectedItem.toString()))?.dayOfYear()?.withMaximumValue()
        }
        if (monthLayout.visibility == View.VISIBLE) {
            period.from = period.from?.withMonthOfYear(Integer.valueOf(spinnerMonth.selectedItem.toString()))?.dayOfMonth()?.withMinimumValue()
            period.to = period.to?.withMonthOfYear(Integer.valueOf(spinnerMonth.selectedItem.toString()))?.dayOfMonth()?.withMaximumValue()
        }
        if (dayLayout.visibility == View.VISIBLE) {
            period.from = period.from?.withDayOfMonth(Integer.valueOf(spinnerDay.selectedItem.toString()))
            period.to = period.to?.withDayOfMonth(Integer.valueOf(spinnerDay.selectedItem.toString()))
        }
    }

    private fun findOr0(array: Array<Int>, value: Int): Int {
        return if (Arrays.asList(*array).contains(value)) Arrays.asList(*array).indexOf(value) else 0
    }

    private fun createSpinnerAdapter(strings: Array<Int>): ArrayAdapter<*> =
            ArrayAdapter(context, android.R.layout.simple_spinner_item, strings).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }

    private fun createSpinnerAdapter(strings: Array<String>): ArrayAdapter<*> =
            ArrayAdapter(context, android.R.layout.simple_spinner_item, strings).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }
}